# Grouping Data - Task

In [2]:
# Import libraries 
import pandas as pd 
import numpy as np 
import os 

In [3]:
# Defining path 
path = r'C:\Users\ale_o\OneDrive\Escritorio\Data Analytics\5. Python\Instacart Basket Analysis 01 - 2023 final'

In [4]:
# Importin file 
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [5]:
df_ords_prods_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   
5                    19.0       196.0                1.0        1.0   
6                    20.0       196.0                1.0        1.0   
7                    14.0       196.0                2.0        1.0   
8                     0.0       196.0                4.0        1.0   
9                    30.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both  
5         Soda        77              7     9.0   both  
6         Soda        77              7     9.0   both  
7         Soda        77              7     9.0   both  
8         Soda        77              7     9.0   both  
9         Soda        77              7     9.0   both

In [6]:
df_ords_prods_merge.shape

(32404859, 14)

#### 2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [7]:
df_ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64


#### 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

The results obtained are more reliable since we have a complete picture of the data. We have information of all the departments and some results are different to the previos calculations made with the subset. 
For example de average for departmet 4 using the subset was 18.25 orders, and using the whole data frame the result is 17.8 
Although we had a good visibility of the data using the dataframe, since the dataframe is very big the results can differ. 

#### 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [8]:
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [9]:
df_ords_prods_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   
5                    19.0       196.0                1.0        1.0   
6                    20.0       196.0                1.0        1.0   
7                    14.0       196.0                2.0        1.0   
8                     0.0       196.0                4.0        1.0   
9                    30.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices _merge  max_order  
0         Soda        77              7     9.0   both         10  
1         Soda        77              7     9.0   both         10  
2         Soda        77              7     9.0   both         10  
3         Soda        77              7     9.0   both         10  
4         Soda        77              7     9.0   both         10  
5         Soda        77              7     9.0   both         10  
6         Soda        77              7     9.0   both         10  
7         Soda        77              7     9.0   both         10  
8         Soda        77              7     9.0   both         10  
9         Soda        77              7     9.0   both         10

In [10]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

#### 5.What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [11]:
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max', 'sum']})

prices                           
                       mean  min      max          sum
loyalty_flag                                          
Loyal customer    10.386336  1.0  99999.0  106814042.2
New customer      13.294670  1.0  99999.0   83011787.2
Regular customer  12.495717  1.0  99999.0  198391693.2

#### 6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [12]:
# creating a red flag based on average price 
df_ords_prods_merge['avg_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.max)

In [13]:
df_ords_prods_merge.head (10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   
5                    19.0       196.0                1.0        1.0   
6                    20.0       196.0                1.0        1.0   
7                    14.0       196.0                2.0        1.0   
8                     0.0       196.0                4.0        1.0   
9                    30.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices _merge  max_order  \
0         Soda        77              7     9.0   both         10   
1         Soda        77              7     9.0   both         10   
2         Soda        77              7     9.0   both         10   
3         Soda        77              7     9.0   both         10   
4         Soda        77              7     9.0   both         10   
5         Soda        77              7     9.0   both         10   
6         Soda        77              7     9.0   both         10   
7         Soda        77              7     9.0   both         10   
8         Soda        77              7     9.0   both         10   
9         Soda        77              7     9.0   both         10   

   loyalty_flag  avg_price  
0  New customer       14.0  
1  New customer       14.0  
2  New customer       14.0  
3  New customer       14.0  
4  New customer       14.0  
5  New customer       14.0  
6  New customer       14.0  
7  New customer       14.0  
8  New customer       14.0  
9  New customer       14.0

In [14]:
# Creating red flag 
df_ords_prods_merge.loc[df_ords_prods_merge['avg_price'] < 10, 'spender_criteria'] = 'Low spender'

In [15]:
df_ords_prods_merge.loc[df_ords_prods_merge['avg_price'] >= 10, 'spender_criteria'] = 'High spender'

In [16]:
df_ords_prods_merge['spender_criteria'].value_counts()

High spender    32372969
Low spender        31890
Name: spender_criteria, dtype: int64

In [17]:
df_ords_prods_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   
5                    19.0       196.0                1.0        1.0   
6                    20.0       196.0                1.0        1.0   
7                    14.0       196.0                2.0        1.0   
8                     0.0       196.0                4.0        1.0   
9                    30.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices _merge  max_order  \
0         Soda        77              7     9.0   both         10   
1         Soda        77              7     9.0   both         10   
2         Soda        77              7     9.0   both         10   
3         Soda        77              7     9.0   both         10   
4         Soda        77              7     9.0   both         10   
5         Soda        77              7     9.0   both         10   
6         Soda        77              7     9.0   both         10   
7         Soda        77              7     9.0   both         10   
8         Soda        77              7     9.0   both         10   
9         Soda        77              7     9.0   both         10   

   loyalty_flag  avg_price spender_criteria  
0  New customer       14.0     High spender  
1  New customer       14.0     High spender  
2  New customer       14.0     High spender  
3  New customer       14.0     High spender  
4  New customer       14.0     High spender  
5  New customer       14.0     High spender  
6  New customer       14.0     High spender  
7  New customer       14.0     High spender  
8  New customer       14.0     High spender  
9  New customer       14.0     High spender

#### 7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”

If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”

If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [18]:
# Defining the median reference
df_ords_prods_merge['frequency_order_median']=df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [19]:
df_ords_prods_merge.groupby('user_id').agg({'days_since_prior_order': ['median']})

days_since_prior_order
                        median
user_id                       
1                         20.5
2                         13.0
3                         10.0
4                         20.0
5                         11.0
...                        ...
206205                    30.0
206206                     3.0
206207                    16.0
206208                     7.0
206209                    22.0

[206209 rows x 1 columns]

In [20]:
df_ords_prods_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   
5                    19.0       196.0                1.0        1.0   
6                    20.0       196.0                1.0        1.0   
7                    14.0       196.0                2.0        1.0   
8                     0.0       196.0                4.0        1.0   
9                    30.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices _merge  max_order  \
0         Soda        77              7     9.0   both         10   
1         Soda        77              7     9.0   both         10   
2         Soda        77              7     9.0   both         10   
3         Soda        77              7     9.0   both         10   
4         Soda        77              7     9.0   both         10   
5         Soda        77              7     9.0   both         10   
6         Soda        77              7     9.0   both         10   
7         Soda        77              7     9.0   both         10   
8         Soda        77              7     9.0   both         10   
9         Soda        77              7     9.0   both         10   

   loyalty_flag  avg_price spender_criteria  frequency_order_median  
0  New customer       14.0     High spender                    20.5  
1  New customer       14.0     High spender                    20.5  
2  New customer       14.0     High spender                    20.5  
3  New customer       14.0     High spender                    20.5  
4  New customer       14.0     High spender                    20.5  
5  New customer       14.0     High spender                    20.5  
6  New customer       14.0     High spender                    20.5  
7  New customer       14.0     High spender                    20.5  
8  New customer       14.0     High spender                    20.5  
9  New customer       14.0     High spender                    20.5

In [21]:
# Creating the frequency customer flag
df_ords_prods_merge.loc[df_ords_prods_merge['frequency_order_median'] > 20, 'frequency'] = 'Non-frequent customer'

In [22]:
df_ords_prods_merge.loc[(df_ords_prods_merge['frequency_order_median'] > 10) & (df_ords_prods_merge['frequency_order_median'] <= 20), 'frequency'] = 'Regular customer'

In [23]:
df_ords_prods_merge.loc[df_ords_prods_merge['frequency_order_median'] <= 10, 'frequency']  = 'Frequent customer'

In [24]:
# Counting customers by frequency 
df_ords_prods_merge['frequency'].value_counts()

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
Name: frequency, dtype: int64

#### 9. Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [25]:
# Exporting dataframe 
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_updated_2.pkl'))

In [26]:
df_ords_prods_merge.tail()

order_id  user_id  order_number  orders_day_of_week  \
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
32404854                 15                     1.0     43553.0   
32404855                 11                     3.0     43553.0   
32404856                  7                     NaN     42338.0   
32404857                  5                    15.0     42338.0   
32404858                 14                     NaN     28818.0   

          add_to_cart_order  reordered                   product_name  \
32404854                2.0        1.0            Orange Energy Shots   
32404855                2.0        1.0            Orange Energy Shots   
32404856                4.0        0.0          Zucchini Chips, Pesto   
32404857               16.0        1.0          Zucchini Chips, Pesto   
32404858                8.0        0.0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices _merge  max_order      loyalty_flag  \
32404854        64              7     3.7   both         31  Regular customer   
32404855        64              7     3.7   both         31  Regular customer   
32404856        50             19     6.9   both          3      New customer   
32404857        50             19     6.9   both          3      New customer   
32404858       130             14    10.3   both         16  Regular customer   

          avg_price spender_criteria  frequency_order_median  \
32404854       24.0     High spender                     8.0   
32404855       24.0     High spender                     8.0   
32404856       14.4     High spender                    15.0   
32404857       14.4     High spender                    15.0   
32404858       24.9     High spender                    13.0   

                  frequency  
32404854  Frequent customer  
32404855  Frequent customer  
32404856   Regular customer  
32404857   Regular customer  
32404858   Regular customer